### PR TITLE
Add home-end keymaps, (as kakoune/vim do)

### PR DESCRIPTION
**Pull request by itzmjauz**
Similar  to my last pr, this extends cursor movement through home and end keymappings.

### CODE REVIEW

File: helix-term/src/keymap.rs [hunk 1]
```diff
@@ -146,7 +146,16 @@ pub fn default() -> Keymaps {
         key!('r') => commands::replace,
 
         key!('^') => commands::move_line_start,
+        KeyEvent {
+            code: KeyCode::Home,
+            modifiers: KeyModifiers::NONE
+        } => commands::move_line_start,
+
```
Comment (pickfire) [inline helix-term/src/keymap.rs:153]: I wonder why we didn't go `gl`.

Comment (itzmjauz) [reply helix-term/src/keymap.rs:153]: I think gl/gi are actually great as well, I feel the entire 'goto mode' of kakoune made a lot of sense in general.

Of course at the end of the day the editor will probably have some configuration files that can be loaded at runtime to adjust key mappings.

I just felt home/end keymaps should be in the default/base configuration.

Comment (archseer) [reply helix-term/src/keymap.rs:153]: I would be open to `gl` `gi` mappings, was just more familiar with the regex-like start/end

Comment (kevinsjoberg) [reply helix-term/src/keymap.rs:153]: I think it makes sense to have some sensible defaults, but then postpone adding too many ways of achieving the same thing. As @itzmjauz said, it's better to let user configuration handle that in the future.

File: helix-term/src/keymap.rs [hunk 2]
```diff
@@ -291,6 +300,14 @@ pub fn default() -> Keymaps {
                 code: KeyCode::Esc,
                 modifiers: KeyModifiers::NONE
             } => commands::exit_select_mode as Command,
+            KeyEvent {
+                code: KeyCode::Home,
+                modifiers: KeyModifiers::NONE
+            } => commands::move_line_start,
```
Comment (itzmjauz) [inline helix-term/src/keymap.rs:306]: Need to use appropriate extend methods here as well

Comment (archseer) [reply helix-term/src/keymap.rs]: Can you add these extend mappings for `^` and `$` too?

Comment (itzmjauz) [reply helix-term/src/keymap.rs]: Say no more ~ on it

**Comment by itzmjauz** [comment]
Undone the merge, for some reason shift!() is not implemented here so it won't build

**Comment by archseer** [comment]
`shift!()` was removed on master, keys should now simply use `key!()` This is so that there's no difference between `Z` and `shift` `Z`

**Comment by itzmjauz** [comment]
> `shift!()` was removed on master, keys should now simply use `key!()` This is so that there's no difference between `Z` and `shift` `Z`

I figured, checked the history, should work now, Might want to rebase/squash this entire thing though

**Comment by pickfire** [comment]
> shift!() was removed on master, keys should now simply use key!() This is so that there's no difference between Z and shift Z

But there is no difference between `Z` and `ctrl!(Z)` though. How do we handle this?

**Comment by archseer** [comment]
Sorry, can you resolve the conflicts?

**Comment by archseer** [comment]
> > shift!() was removed on master, keys should now simply use key!() This is so that there's no difference between Z and shift Z
> 
> But there is no difference between `ctrl!(z)` and `ctrl!(Z)` though (for terminal). How do we handle this?

Let's open an issue for this, I was wondering how kakoune resolved this. The canonicalization change was made after some testing on windows where certain terminals wouldn't report a shift modifier with ":", thus not being able to enter command mode.

**Comment by pickfire** [comment]
> Let's open an issue for this, I was wondering how kakoune resolved this. The canonicalization change was made after some testing on windows where certain terminals wouldn't report a shift modifier with ":", thus not being able to enter command mode.

I have an idea for these, just enumerate all the possible items in an enum. But let's do that after this.

**Comment by itzmjauz** [comment]
> Sorry, can you resolve the conflicts?

Sure, I'm not sure what functionality you want now though,
Rather than extend ^ and $, the bindings are now completely removed.

Do you want only gh/gl + Home-end ?

File: helix-term/src/keymap.rs [hunk 3]
```diff
@@ -309,13 +318,23 @@ pub fn default() -> Keymaps {
...
+            key!('^') => commands::extend_line_start,
+            KeyEvent {
+                code: KeyCode::Home,
+                modifiers: KeyModifiers::NONE
+            } => commands::extend_line_start,
+            key!('$') => commands::extend_line_end,
+            KeyEvent {
+                code: KeyCode::End,
+                modifiers: KeyModifiers::NONE
+            } => commands::extend_line_end,
```
Comment (pickfire) [inline helix-term/src/keymap.rs:333]: Can you just merge them with `|`?

**Comment by pickfire** [comment]
> Sure, I'm not sure what functionality you want now though,

Sorry for not being clear.

Remove <kbd>^</kbd> and <kbd>$</kbd>.
Add <kbd>g h</kbd>, <kbd>g l</kbd>, <kbd>Home</kbd>, <kbd>End</kbd>.

**Comment by itzmjauz** [comment]
> > Sure, I'm not sure what functionality you want now though,
> 
> Sorry for not being clear.
> 
> Remove ^ and $.
> Add g h, g l, Home, End.

Since goto is its own mode, it directly conflicts with extend mode. 
You cannot use any goto commands to extend a selection because of this. 
Home and End allow this at the moment, but in extend-mode one cannot do 'g-l' to extend the selection to the end of the line.
I'm not sure if this is intended behaviour.

File: helix-term/src/keymap.rs (hunk 3 above)
Comment (itzmjauz) [reply helix-term/src/keymap.rs]: No point since they're deleted now, but yes, why not; I just wanted to be consistent with the rest of the code.

**Comment by archseer** [comment]
Hmm, seems like there's still some conflicts with master?

Yeah goto doesn't work with extend yet, but I'll fix that later :)

**Comment by itzmjauz** [comment]
> Hmm, seems like there's still some conflicts with master?
> 
> Yeah goto doesn't work with extend yet, but I'll fix that later :)

I'm not seeing any? 
I can pull/build my branch into master fine locally.

If this is getting too messy I could go for a new PR;

**Comment by archseer** [comment]
Looks like I needed to "Squash and merge". Thanks!

**Pull request merged**